src/full/Agda/TypeChecking/Patterns/Match.hs

```diff
@@ -55,24 +57,25 @@ buildSubstitution :: (DeBruijn a)
 buildSubstitution err n vs = parallelS $ map unArg $ matchedArgs err n vs
 
 
--- 'mappend' is UNUSED.
---
--- instance Monoid (Match a) where
---     mempty = Yes mempty []
+instance Semigroup (Match a) where
+    Yes s us   <> Yes s' vs        = Yes (s <> s') (us <> vs)
+    Yes _ _    <> No               = No
+    Yes _ _    <> DontKnow m       = DontKnow m
+    No         <> DontKnow m       = DontKnow m -- Issue 2964: We need to be conservative wrt to splitting order
+    No         <> _                = No
 
---     Yes s us   `mappend` Yes s' vs        = Yes (s `mappend` s') (us ++ vs)
---     Yes _ _    `mappend` No               = No
---     Yes _ _    `mappend` DontKnow m       = DontKnow m
---     No         `mappend` _                = No
+    -- @NotBlocked (StuckOn e)@ means blocked by a variable.
+    -- In this case, no instantiation of
+    -- meta-variables will make progress.
+    DontKnow b <> DontKnow b'      = DontKnow $ b <> b'
 
---     -- @NotBlocked (StuckOn e)@ means blocked by a variable.
---     -- In this case, no instantiation of
---     -- meta-variables will make progress.
---     DontKnow b `mappend` DontKnow b'      = DontKnow $ b `mappend` b'
+    -- One could imagine DontKnow _ <> No = No, but would break the
+    -- equivalence to case-trees.
+    DontKnow m <> _                = DontKnow m
```

Comment:
This pattern match could be reordered for clarity: `DontKnow` is dominant, so put the cases involving `DontKnow` on top (3 cases).  Next in the pecking order is `No` (2 cases).  Last is `Yes` (1 case).